pkg/boot/linux.go

```diff
@@ -191,22 +215,22 @@ func copyToFileIfNotRegular(r io.ReaderAt, verbose bool) (*os.File, error) {
 	// conforming to os.File. We then can derive file descriptor, and the
 	// name.
 	if f, ok := r.(*os.File); ok {
-		if fi, err := f.Stat(); err == nil && fi.Mode().IsRegular() {
-			if r, _ := mount.IsTmpRamfs(f.Name()); r {
-				// Check if original file is opened for write. Perform copy to
-				// get a read only version in that case, than directly return
-				// as kexec load would fail on a file opened for write.
-				wr := unix.O_RDWR | unix.O_WRONLY
-				if am, err := unix.FcntlInt(f.Fd(), unix.F_GETFL, 0); err == nil && am&wr == 0 {
-					return f, nil
-				}
-				// Original file is either opened for write, or it failed to
-				// check (possibly, current kernel is too old and not supporting
-				// the sys call cmd)
-			}
-			// Original file is neither on a tmpfs, nor a ramfs.
+		if isTmpfsReadOnlyFile(f) {
+			return f, nil
+		}
+	}
+
+	// For boot entries whose image is lazily downloaded, try booting the
+	// backend file directly if possible.
+	if lor, ok := r.(*uio.LazyOpenerAt); ok {
+		// It is lazy, so read one byte to make sure backend file
+		// is created.
+		if err := uio.ReadOneByte(r); err != nil {
+			return nil, err
+		}
+		if isTmpfsReadOnlyFile(lor.File()) {
```

Comment:
i think calling File() should open it

Reply:
not sure if I follow, anything worth changing ? -- File() does return a read only `*os.File`, which is only non-nil, after 
 first `ReadAt` call.

Reply:
actually, I do need it, in `pkg/boot/initrd.go`,

`names = append(names, stringer(initrd))`, where `initrd` is a io.Reader.

Reply:
the premise of all the lazy functions, like ReadAt, is that when you call it the first time it opens the file. File() should be like that. Rather than having to read one byte, you just call File() and it opens the file and returns it

Reply:
any specific suggestions for how to write differently ? 

My view is that this is still LazyReaderAt, that fetches http body lazily. Not a LazyFile opening a path.  returning a *os.File is a by-product, which is only non-nil, after first ReadAt call. I am hesitant to make reading one byte as part of File() call for same reason.  File() here simply returns an internal reference, `f`.